fixtures/_fixtures/monkeypatch.py

```diff
@@ -23,21 +23,91 @@
 from fixtures import Fixture
 
 
-def _setattr(obj, name, value):
-    """Handle some corner cases when calling setattr.
+class PartialClassMethod(object):
+    """Emulates the classmethod descriptor, but works on bound methods.
+
+    This wraps a function so that it acts like a classmethod on the calling
+    class. If the wrapped function is already bound to a class then the passed
+    parameters look like foo(boundcls_or_self, callingcls, *args, **kwargs).
 
-    setattr transforms a function into instancemethod, so where appropriate
-    value needs to be wrapped with staticmethod().
     """
+    def __init__(self, f):
+        self.f = f
+
+    def __call__(self, *args, **kwargs):
+        # Called when this is wrapped in another descriptor so __get__ is not
+        # used.
+        return self.f(*args, **kwargs)
+
+    def __get__(self, obj, objtype=None):
+        if objtype is None:
+            objtype = type(obj)
+
+        def newfunc(*args, **kwargs):
+            return self.f(objtype, *args, **kwargs)
+        return newfunc
```

Comment:
I think here you'd be better off using partial() directly:
return partial(self.f, objtype)

Its a bit of a shame that we need this - but AFAICT the issue is that a boundmethod instance still has `__get__` on it?

```
class patch(object):
   @classmethod
   def kls(cls, tgtclass, one): pass

patch.kls(1,2)
patch.kls # <- bound method - an instancemethod instance
patch().kls # - equally.
```
We got those from `__get__`.

where t is an instance of a class:
```
>>> t.z = patch.kls
>>> t.z(1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: kls() takes exactly 3 arguments (2 given)
>>> t.z(1,2)
>>> type(t).z = patch.kls
>>> t.z()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: kls() takes exactly 3 arguments (1 given)
```
The first case above we're calling a callable directly, and the class of t won't be accessible.
In the second case, our descriptor is being unwrapped by t, so the class of patch won't be accessible - and we're breaking the model of being a simple callable.

But, if we wrap them with -any- proxy at all, we prevent that, no?
```
>>> t.x = classmethod(patch.kls)
>>> t.x(1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'classmethod' object is not callable
```
^- this is because we assigned to the instance, so no descriptor logic gets invoked.

```
>>> type(t).x = classmethod(patch.kls)
>>> t.x(1)
>>>
```
OTOH, that means we're now assigning to the type when a user wants to patch a single instance.

But, AFAICT your code depends on descriptor logic too operate (assigning a bound method to an instance attribute doesn't need to do anything special because `__get__` won't be called). So - can't we just wrap directly in classmethod etc?